ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -73,46 +114,72 @@ static double started_tracing_at_microseconds = 0;
 static int64_t process_id = 0;
 static VALUE gc_tracepoint = Qnil;
 
+static VALUE tracing_init_local_storage(VALUE, VALUE);
+static VALUE tracing_start(VALUE _self, VALUE output_path);
+static VALUE tracing_stop(VALUE _self);
+static double timestamp_microseconds(void);
+static void set_native_thread_id(thread_local_state *);
+static void render_event(thread_local_state *, const char *event_name);
+static void on_thread_event(rb_event_flag_t event, const rb_internal_thread_event_data_t *_unused1, void *_unused2);
+static void on_gc_event(VALUE tpval, void *_unused1);
+static VALUE mark_sleeping(VALUE _self);
+
 void Init_gvl_tracing_native_extension(void) {
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET // 3.3+
+  thread_storage_key = rb_internal_thread_specific_key_create();
+#endif
+
   rb_global_variable(&gc_tracepoint);
 
   VALUE gvl_tracing_module = rb_define_module("GvlTracing");
 
+  rb_define_singleton_method(gvl_tracing_module, "_init_local_storage", tracing_init_local_storage, 1);
   rb_define_singleton_method(gvl_tracing_module, "_start", tracing_start, 1);
   rb_define_singleton_method(gvl_tracing_module, "_stop", tracing_stop, 0);
   rb_define_singleton_method(gvl_tracing_module, "mark_sleeping", mark_sleeping, 0);
 }
 
-static inline void initialize_thread_id(void) {
-  current_thread_seen = true;
-  current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
-  set_native_thread_id();
+static inline void initialize_thread_id(thread_local_state *state) {
+  state->current_thread_seen = true;
+  state->current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
+  set_native_thread_id(state);
 }
 
-static inline void render_thread_metadata(void) {
+static inline void render_thread_metadata(thread_local_state *state) {
   char native_thread_name_buffer[64] = "(unnamed)";
 
   #ifdef HAVE_PTHREAD_GETNAME_NP
     pthread_getname_np(pthread_self(), native_thread_name_buffer, sizeof(native_thread_name_buffer));
   #endif
 
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET
+  uint64_t thread_id = (uint64_t)state->thread;
```

Comment:
Yeah, we haven't concerned ourselves much with names in the resulting JSON. I'd rather go baby step to make the things easier to review. Having the valid timeline is more important than having the name, so we did this first.

Having the actual Ruby thread in there should make it easy to have the correct names later. Anyway the whole `native_thread_id` thing should be ignored because of MaNy.

Reply:
Ideally I'd still want to have the native thread id when possible (e.g. when not using MaNy on the main ractor) as I think it's a useful info as you can use it to correlate with other native-level tools but totally fine to ignore that and simplify for now :)

Reply:
I think you could collect it even with MaNy, it just it can't be the thread identifier anymore.

Reply:
I completely forgot - I tried using `VALUE` initially, based on your recommendation elsewhere @byroot. I switched to `object_id` because `VALUE` is too large - it is invalid in the perfetto ui. If I try importing the file this PR generates, none of the threads show up. Annoyingly, `tid` is defined as an `int32` and so perfetto completely ignores/rejects every thread written using `VALUE`.

Reply:
I guess perhaps we could go back to the sequential thread id to make perfetto happy, and then we could have all relevant ids as part of the thread name string maybe?

Reply:
I've hacked something where I xor the 32 higher bits with the lower 32 ones of the VALUE, that way `tid` is 32 bits (which I assume is what's expected by Perfetto UI since TIDs are PIDs and `pid_t` is 32 bits).

Now the threads are visible, even if they've lost their name in the thread_name_with_extension example.